In [321]:
# Imports and pip installations (if needed)
import pandas as pd
pd.set_option('display.max_rows', 100)     # Show all rows
pd.set_option('display.max_columns', 25)  # Show all columns
pd.set_option('display.expand_frame_repr', False)  # Prevent line breaks in the output

# Part 1: Load the dataset

In [322]:
# Load the given datasets
categorical_data = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical_data = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
categorical_data

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 
<br>The data is a collection of 24 categorical and numerical features which in some way can influence whether someone has or doesn't have chronic kidney disease. Categorical data can be expressed as a part of a certain amount of categories, whereas numerical data can be expressed as a number. Since each of the features can be used as a "mini-predictor" for the disease, they can all be combined together as factors and used as inputs into the machine learning model to determine whether somebody has chronic kidney disease. There is one label in the dataset, which is essentially what the model should output as a result and this is the class column. Here it is labeled either "notckd" or "ckd" representing whether or not each of the subjects has chronic kidney disease.
<br>Three examples of numerical data include: age, blood pressure and sodium levels.
<br>Three examples of categorical data include: anemia, coronary artery disease, hypertension.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [323]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates_numerical = sum(numerical_data.duplicated(subset=['unique_id'], keep=False))
print("Total duplicate rows in numerical: " , duplicates_numerical)

# Drop these duplicate rows
numerical_data = numerical_data.drop_duplicates(subset=['unique_id'], keep='first')

# USE TO CHECK IF WORKING
# duplicates_numerical = sum(numerical_data.duplicated(subset=['unique_id'], keep=False))
# print(duplicates_numerical)


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates_categorical = sum(categorical_data.duplicated(subset=['unique_id'], keep=False))
print("Total duplicate rows in categorical: ", duplicates_categorical)
categorical_data = categorical_data.drop_duplicates(subset=['unique_id'], keep='first')

# USE TO CHECK IF WORKING
# duplicates_categorical = sum(categorical_data.duplicated(subset=['unique_id'], keep=False))
# print(duplicates_categorical)

# PRINTED TOTAL AMOUNT OF DUPLICATE ROWS IN EACH OF THE DATASETS. 
# MEANING IF 3 ROWS SHARE A UNIQUE_ID, THEN 3 WILL BE ADDED TO THIS DUPLICATE ROW COUNT INSTEAD OF JUST 2.

Total duplicate rows in numerical:  59
Total duplicate rows in categorical:  59


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
<br> All unique_IDs are NOT present in both datasets. This is likely because both the categorical and numerical data might not have been collected for that same unique_ID (person). This means there is data missing for that subject. Also since I did a join on unique_id, it joins elements from both datasets which share a unique_id. Therefore, if a unique_id is present in one of the datasets, but not in another, then it will form its own row in the merged table where any of the fields not specified with present as NaN or null with missing values. This it doesn't have a corresponding unique_id in the other table, so any of the necessary data from there will be "missing" and therefore NaN.

In [324]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_data = pd.merge(how='outer', on='unique_id', left=categorical_data, right=numerical_data)

#Print the combined dataset
merged_data

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd,54.0,60.0,125.0,21.0,1.3,137.0,3.4,15.0,46.0,NaN,NaN
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd,35.0,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,9400.0,6.1
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd,65.0,70.0,139.0,29.0,1.0,NaN,NaN,10.5,32.0,NaN,NaN
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd,5.0,NaN,NaN,16.0,0.7,138.0,3.2,8.1,NaN,NaN,NaN
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,42.0,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,69.0,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
396,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,43.0,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9
397,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd,40.0,80.0,76.0,162.0,9.6,141.0,4.9,7.6,24.0,3800.0,2.8
398,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,47.0,60.0,117.0,22.0,1.2,138.0,3.5,13.0,45.0,5200.0,5.6


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [325]:
# Calculate the percentage of rows that contain atleast one missing value
num_missing = merged_data.isnull().any(axis=1).sum()
num = merged_data.shape[0]
percent = (num_missing / num) * 100

# Print %
print(f'{round(percent, 3)}%')

# Drop these rows from the dataset
merged_data = merged_data.dropna()

# Print the Dataset
merged_data

61.75%


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,42.0,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
5,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
7,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,73.0,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7
8,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,50.0,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0
9,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,poor,no,no,ckd,59.0,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
392,996076,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2
394,397388,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,good,no,no,ckd,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2
395,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,69.0,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
396,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,43.0,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9


## Part 3.4: Sort the dataset according to the Labels

In [326]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged_data_sorted = merged_data.sort_values(by='Target', ascending=True)
merged_data_sorted_reset = merged_data_sorted.reset_index(drop=True)

# Print the dataset
merged_data_sorted_reset


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,534245,2.0,0.0,normal,abnormal,present,notpresent,no,no,no,good,no,no,ckd,45.0,70.0,117.0,52.0,2.2,136.0,3.8,10.0,30.0,19100.0,3.7
1,349892,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1
2,343710,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2
3,102677,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1
4,514721,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,no,no,ckd,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0
149,530138,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0
150,261137,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,5800.0,6.2
151,322882,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,33.0,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,6200.0,4.5


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [327]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
merged_data_dummies = pd.get_dummies(merged_data_sorted_reset, columns=categorical_columns, dtype=int)

# Print the dataset
merged_data_dummies

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,534245,2.0,0.0,45.0,70.0,117.0,52.0,2.2,136.0,3.8,10.0,30.0,...,1,0,1,0,1,0,1,0,1,0,1,0
1,349892,1.0,3.0,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,0,1,0,1,1,0,1,0,1,0,1,0
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,0,1,1,0
3,102677,3.0,0.0,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,1,0,1,0,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,...,1,0,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,...,1,0,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,...,1,0,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,...,1,0,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,...,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

<br> It would be equivalent to have a single column for a binary variable. This is because the values are mutually exclusive for each of the binary variables, it can be True or False, but never both. Therefore, it can be more efficiently represented in one column instead of two. However, if a categorical variable is not binary and can take on more than 2 values, the behavior with the model would be slightly different as the data would be presented in a different way. If it were true that only one of those values in the categorical column could be true at once, then it would be valid if the data for that variable is all represented in the same column. However, if it possible for multiple values in the same categorical column to be true at once, then it needs to be split into separate columns. Like with the sandwich condiments, there needs to be 5 separate columns for each of the binary variables. This is because, multiple condiments can be on a sandwich at the same time. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [328]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    mean = merged_data_dummies[column].mean()
    std_dev = merged_data_dummies[column].std()
    
    lower_bound = mean - (3 * std_dev)
    upper_bound = mean + (3 * std_dev)
    
    merged_data_dummies = merged_data_dummies[(merged_data_dummies[column] >= lower_bound) & (merged_data_dummies[column] <= upper_bound)]
# Print the dataset
df = merged_data_dummies
df


,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,...,1,0,1,0,1,0,1,0,1,0,1,0
7,124304,3.0,4.0,69.0,70.0,264.0,87.0,2.7,130.0,4.0,12.5,37.0,...,0,1,0,1,1,0,0,1,1,0,1,0
8,980291,3.0,0.0,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,0,1,1,0,1,0,1,0,1,0,1,0
10,615697,4.0,0.0,6.0,60.0,94.0,67.0,1.0,135.0,4.9,9.9,30.0,...,1,0,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,...,1,0,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,...,1,0,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,...,1,0,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,...,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [329]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler

# Creating a standard scaler instance
scaler = MinMaxScaler()

# Fitting and transforming the data
df_standardized = scaler.fit_transform(df)

# Converting back to DataFrame
df_standardized = pd.DataFrame(df_standardized, columns=df.columns)

# Print the dataset
df_standardized

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.264544,0.50,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,1.0,0.0
1,0.457337,0.50,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
2,0.017193,0.75,1.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.379310,0.475248,0.451613,...,0.0,1.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
3,0.982205,0.75,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.620690,0.485149,0.516129,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
4,0.571173,1.00,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,...,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,0.428520,0.00,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.724138,0.980198,0.903226,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
129,0.474717,0.00,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.379310,0.653465,0.580645,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
130,0.171454,0.00,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.344828,0.693069,0.645161,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
131,0.241064,0.00,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.344828,0.534653,0.709677,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: 
I removed the unique_id because it doesn't provide any input for the machine learning model to predict chronic kidney disease. The only use for the unique_id column is for data preprocessing reasons, this way we can attribute each line of data to a unique subject. Once our data preprocessing is complete it is best to get rid of that column since it is not a variable affecting chronic kidney disease.

In [330]:
#Remove that column

#Remove the unique_id and index because it is not needed for machine learning interpretation
final_df = df_standardized.drop(columns=['unique_id'])

# Print the dataset
final_df

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.50,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,1.0,0.0
1,0.50,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
2,0.75,1.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.379310,0.475248,0.451613,0.427419,...,0.0,1.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
3,0.75,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.620690,0.485149,0.516129,0.290323,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
4,1.00,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,1.000000,...,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,0.00,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.724138,0.980198,0.903226,0.177419,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
129,0.00,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.379310,0.653465,0.580645,0.233871,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
130,0.00,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.344828,0.693069,0.645161,0.120968,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
131,0.00,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.344828,0.534653,0.709677,0.153226,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [331]:
# Export the dataframe to a new csv file
final_df.to_csv('cleaned_ckd_data.csv', index=False)  

# Export the dataframe to a new json file
final_df.to_json('cleaned_ckd_data.json', index=False, orient='split')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [332]:
final_df.head(15)

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.50,0.00,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,1.0,0.0
1,0.50,0.00,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
2,0.75,1.00,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.379310,0.475248,0.451613,0.427419,...,0.0,1.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
3,0.75,0.00,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.620690,0.485149,0.516129,0.290323,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
4,1.00,0.00,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,0.225806,1.000000,...,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0
5,0.75,0.25,0.662162,0.50,0.618026,0.411765,0.432099,0.566667,0.689655,0.316832,0.354839,0.250000,...,0.0,1.0,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
6,0.25,0.00,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
7,1.00,0.50,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.586207,0.019802,0.096774,0.258065,...,0.0,1.0,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
8,1.00,0.00,0.567568,0.50,0.107296,1.000000,0.901235,0.533333,0.310345,0.207921,0.290323,0.209677,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
9,1.00,0.75,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.344828,0.168317,0.161290,0.580645,...,0.0,1.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0


### Paste the markdown table here

markdown

| al    | su    | age      | bp    | bgr      | bu        | sc        | sod       | pot       | hemo     | pcv      | wbcc     | ... | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-------|-------|----------|-------|----------|-----------|-----------|-----------|-----------|----------|----------|----------|-----|-------|--------|--------|---------|------------|-------------|-------|--------|--------|---------|------------|----------------|
| 0.50  | 0.00  | 0.743243 | 0.50  | 0.442060 | 0.901961  | 0.432099  | 0.500000  | 0.793103  | 0.000000 | 0.032258 | 0.395161 | ... | 0.0   | 1.0    | 0.0    | 1.0     | 0.0        | 1.0         | 0.0   | 1.0    | 0.0    | 1.0     | 1.0        | 0.0            |
| 0.50  | 0.00  | 0.675676 | 0.75  | 0.253219 | 0.633987  | 0.777778  | 0.366667  | 0.655172  | 0.138614 | 0.193548 | 0.169355 | ... | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 0.75  | 1.00  | 0.851351 | 0.25  | 0.832618 | 0.503268  | 0.283951  | 0.333333  | 0.379310  | 0.475248 | 0.451613 | 0.427419 | ... | 0.0   | 1.0    | 0.0    | 1.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 0.75  | 0.00  | 0.756757 | 0.25  | 0.223176 | 0.209150  | 0.160494  | 0.533333  | 0.620690  | 0.485149 | 0.516129 | 0.290323 | ... | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 1.00  | 0.00  | 0.000000 | 0.00  | 0.103004 | 0.372549  | 0.074074  | 0.500000  | 0.689655  | 0.217822 | 0.225806 | 1.000000 | ... | 1.0   | 0.0    | 1.0    | 0.0     | 0.0        | 1.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 0.75  | 0.25  | 0.662162 | 0.50  | 0.618026 | 0.411765  | 0.432099  | 0.566667  | 0.689655  | 0.316832 | 0.354839 | 0.250000 | ... | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 0.25  | 0.00  | 0.540541 | 0.00  | 0.399142 | 0.535948  | 0.358025  | 0.700000  | 0.379310  | 0.207921 | 0.161290 | 0.830645 | ... | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 1.00  | 0.50  | 0.783784 | 1.00  | 0.399142 | 0.287582  | 0.839506  | 0.666667  | 0.586207  | 0.019802 | 0.096774 | 0.258065 | ... | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 1.00  | 0.00  | 0.567568 | 0.50  | 0.107296 | 1.000000  | 0.901235  | 0.533333  | 0.310345  | 0.207921 | 0.290323 | 0.209677 | ... | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 0.0    | 1.0     | 1.0        | 0.0            |
| 1.00  | 0.75  | 0.851351 | 0.25  | 0.618026 | 0.562092  | 0.728395  | 0.000000  | 0.344828  | 0.168317 | 0.161290 | 0.580645 | ... | 0.0   | 1.0    | 0.0    | 1.0     | 1.0        | 0.0         | 0.0   | 1.0    | 0.0    | 1.0     | 1.0        | 0.0            |
| 0.75  | 0.50  | 0.905405 | 1.00  | 0.965665 | 0.522876  | 0.641975  | 0.666667  | 0.000000  | 0.148515 | 0.225806 | 0.217742 | ... | 0.0   | 1.0    | 0.0    | 1.0     | 0.0        | 1.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 1.00  | 0.00  | 0.202703 | 0.75  | 0.158798 | 0.196078  | 0.160494  | 0.166667  | 0.206897  | 0.059406 | 0.000000 | 0.653226 | ... | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 0.0    | 1.0     | 1.0        | 0.0            |
| 1.00  | 0.25  | 0.783784 | 0.00  | 0.725322 | 0.313725  | 0.481481  | 0.566667  | 0.862069  | 0.178218 | 0.193548 | 0.258065 | ... | 0.0   | 1.0    | 1.0    | 0.0     | 0.0        | 1.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 1.00  | 0.25  | 0.675676 | 0.25  | 0.600858 | 0.104575  | 0.160494  | 0.533333  | 0.310345  | 0.831683 | 0.935484 | 0.661290 | ... | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0         | 1.0   | 0.0    | 1.0    | 0.0     | 1.0        | 0.0            |
| 1.00  | 0.00  | 0.567568 | 0.50  | 0.270386 | 0.843137  | 1.000000  | 0.400000  | 0.896552  | 0.257426 | 0.419355 | 0.153226 | ... | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0         | 0.0   | 1.0    | 1.0    | 0.0     | 1.0        | 0.0            |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?
<br>On first inspection, 383gpt properly converted the terminal output into a markdown table. It also created a visually appealing image once pasted into the jupyter notebook markdown. In general though, LLMs can be prone to making mistakes in these kinds of data conversions. There are many ways to mitigate these risks in data conversion. First of all, the prompt should be crystal clear about what the user wants the LLM to accomplish. This eliminates any of the main confusion for when the LLM is evaluating the prompt. Another important method for minimizing errors in this process is processing the data in chunks instead of pasting a large dataset all at once. Giving it smaller loads to handle at a time will reduce the amount of possible errors it could result in at once. It will also be able to find errors while scanning through, which is important since we are almost blindly trusting an LLM to conduct the data conversion. Lastly, we could develop an automated tool which can check if the LLM properly converted the data. At this point though this would become inefficient as we'd have to create multiple tools to check an LLMs answer when we could put that effort into having a tool do the data conversion for us instead.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [333]:
# Code to rename all the columns in the dataset

column_mappings = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class',  # Assuming there isn't a "class" column in the data
    'dm_no': 'diabetes mellitus no',
    'dm_yes': 'diabetes mellitus yes',
    'cad_no': 'coronary artery disease no',
    'cad_yes': 'coronary artery disease yes',
    'appet_good': 'appetite good',
    'appet_poor': 'appetite poor',
    'pe_no': 'pedal edema no',
    'pe_yes': 'pedal edema yes',
    'ane_no': 'anemia no',
    'ane_yes': 'anemia yes',
    'Target_ckd': 'target chronic kidney disease',
    'Target_notckd': 'target not chronic kidney disease',
}

# Rename the columns using the mapping
final_df.rename(columns=column_mappings, inplace=True)

# Display the renamed DataFrame
print(final_df)

     albumin  sugar       age  blood pressure  blood glucose random  blood urea  serum creatinine    sodium  potassium  hemoglobin  packed cell volume      wbcc  ...  diabetes mellitus no  diabetes mellitus yes  coronary artery disease no  coronary artery disease yes  appetite good  appetite poor  pedal edema no  pedal edema yes  anemia no  anemia yes  target chronic kidney disease  target not chronic kidney disease
0       0.50    0.0  0.743243            0.50              0.442060    0.901961          0.432099  0.500000   0.793103    0.000000            0.032258  0.395161  ...                   0.0                    1.0                         0.0                          1.0            0.0            1.0             0.0              1.0        0.0         1.0                            1.0                                0.0
1       0.50    0.0  0.675676            0.75              0.253219    0.633987          0.777778  0.366667   0.655172    0.138614            0.193548  0.169355

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [334]:
# Converted SQL to Pandas code
result = final_df.groupby('target chronic kidney disease').size().reset_index(name='count')
print(result)

   target chronic kidney disease  count
0                            0.0    111
1                            1.0     22


*Describe what the above code does here*

The SQL query is counting the number of subjects which have chronic kidney disease and also the amount of don't have chronic kidney disease. The query groups by 'Target' and since it is a binary variable there are only the 0 and 1 categories. So it will have two rows in the output table with possible values for Target as 0 and 1, and in the second column the number of each of those in the data table. 